napari/utils/transforms/transform_utils.py

```diff
@@ -84,18 +14,83 @@ def compose_linear_matrix(rotate, scale, shear) -> np.array:
             )
         )
     if np.array(shear).ndim == 1:
-        shear_mat = expand_upper_triangular(shear)
-    else:
-        shear_mat = np.array(shear)
+        return expand_upper_triangular(shear)
 
-    # Check the dimensionality of the transforms and pad as needed
-    n_shear = shear_mat.shape[0]
-    ndim = max(n_scale, n_rotate, n_shear)
+    if not is_matrix_triangular(shear):
+        raise ValueError(
+            trans._(
+                'Only upper triangular or lower triangular matrices are accepted for shear, got {shear}. For other matrices, set the affine_matrix or linear_matrix directly.',
+                deferred=True,
+                shear=shear,
+            )
+        )
+    return np.array(shear)
+
+
+def coerce_rotate(rotate):
+    if np.isscalar(rotate):
+        return _make_2d_rotation(rotate)
+    elif np.array(rotate).ndim == 1 and len(rotate) == 3:
+        return _make_3d_rotation(*rotate)
+    return np.array(rotate)
+
+
+""" Makes a 2D rotation matrix from an angle in degrees. """
+
+
+def _make_2d_rotation(theta_degrees):
+    theta = np.deg2rad(theta_degrees)
+    return np.array(
+        [[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]]
+    )
+
+
+""" Makes a 3D rotation matrix from roll, pitch, and yaw in degrees.
+
+For more details, see: https://en.wikipedia.org/wiki/Rotation_matrix#General_rotations
+"""
+
+
+def _make_3d_rotation(alpha_deg, beta_deg, gamma_deg):
+    alpha = np.deg2rad(alpha_deg)
+    beta = np.deg2rad(beta_deg)
+    gamma = np.deg2rad(gamma_deg)
+    R_alpha = np.array(
+        [
+            [np.cos(alpha), np.sin(alpha), 0],
+            [-np.sin(alpha), np.cos(alpha), 0],
+            [0, 0, 1],
+        ]
+    )
+    R_beta = np.array(
+        [
+            [np.cos(beta), 0, np.sin(beta)],
+            [0, 1, 0],
+            [-np.sin(beta), 0, np.cos(beta)],
+        ]
+    )
+    R_gamma = np.array(
+        [
+            [1, 0, 0],
+            [0, np.cos(gamma), -np.sin(gamma)],
+            [0, np.sin(gamma), np.cos(gamma)],
+        ]
+    )
+    return R_alpha @ R_beta @ R_gamma
```

Comment:
just taking a peek - if this is something that will be called a lot it might be worth precalculating the trig calls

Reply:
This is existing code (I just refactored that into this function to help myself organize it) that wasn't called too often before this PR (when any layer's transform parameters were updated), and should be called even less after this (only when the layer's rotate parameter is changed).

But it may still be worth doing some pre-calculation. Did you just mean storing `cos_alpha = np.cos(alpha)` and `sin_alpha = np.sin(alpha)` (likewise for `beta` and `gamma`) or something fancier?